core/data_manager.py

```diff
@@ -195,6 +195,26 @@ def add_rule(self, rule: dict, user_id: str) -> bool:
         except Error as e:
             print(e)
             return False
+
+    def edit_rule(self, rule: dict, user_id: str, rule_id: str) -> bool:
+        try:
+            cur = self.db_conn.cursor()
+            cur.execute('''INSERT INTO rules (id, "key", name, pattern, constraints, rewrite, actions, user_id) 
```

Comment:
Won't this be adding a new rule into the `rules` table instead of updating the existing rule with new values?

Reply:
INSERT INTO rules (id, "key", name, pattern, constraints, rewrite, actions, user_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?) ON CONFLICT (id) DO UPDATE SET name=excluded.name, pattern=excluded.pattern, constraints=excluded.constraints, rewrite=excluded.rewrite, actions=excluded.actions
ON CONFLICT will check if there is a rule with the same id, and if there is so, DO UPDATE SET will modify the name, pattern, constraints, rewrite, and actions, leaving the id untouched.